mathjax3-ts/output/chtml/FontData.ts

```diff
@@ -0,0 +1,328 @@
+    protected delimiters: DelimiterMap = {};
+    protected sizeVariants: string[];
+
+    /*
+     * The actual font parameters for this font
+     */
+    public params: FontParameters;
+
+    /*
+     * Copies the data from the defaults to the instance
+     *
+     * @constructor
+     */
+    constructor() {
+        let CLASS = (this.constructor as typeof FontData);
+        this.params = Object.assign({}, CLASS.defaultParams);
+        this.sizeVariants = CLASS.defaultSizeVariants.slice(0);
+        this.createVariants(CLASS.defaultVariants);
+        this.defineDelimiters(CLASS.defaultDelimiters);
+        for (const name of Object.keys(CLASS.defaultChars)) {
+            this.defineChars(name, CLASS.defaultChars[name]);
+        }
+    }
+
+    /*
+     * Creates the data structure for a variant (an object with prototype chain
+     *   that includes a copy of the linked variant, and then the inherit variant chain.
+     *   The linked copy is updated automatically when the link variant is modified.
+     *   (The idea is to be able to have something like bold-italic inherit from both
+     *   bold and intalic by having the prototype chain include a copy of bold plus
+     *   the full italic chain.  So if something is not defined explicitly in bold-italic,
+     *   it defaults first to a bold version, than an italic version, then the normal
+     *   version, which is in the italic prototype chain.)
+     *
+     * @param{string} name     The new variant to create
+     * @param{string} inherit  The variant to use if a character is not in this one
+     * @param{string} link     A variant to search before the inherit one (but only
+     *                           its top-level object).
+     */
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
+            this.variant[link].linked.push(variant.chars);
+            variant.chars = Object.create(variant.chars);
+        }
+        this.variant[name] = variant;
+    }
+
+    /*
+     * Create a collection of variants
+     *
+     * @param{string[][]} variants  Array of [name, inherit?, link?] values for
+     *                              the variants to define
+     */
+    public createVariants(variants: string[][]) {
+        for (const variant of variants) {
+            this.createVariant.apply(this, variant);
+        }
+    }
+
+    /*
+     * Defines new character data in a given variant
+     *
+     * @param{srtring} name   The variant for these characters
+     * @param{CharMap} chars  The characters to define
+     */
+    public defineChars(name: string, chars: CharMap) {
+        let variant = this.variant[name];
+        Object.assign(variant.chars, chars);
```

Comment:
dito

Reply:
Again, using spread would destroy the prototype chain that is part of `variant.chars`.  Also, since `variant.chars` may not be empty, you would have to use something like
```
variant.chars = {...variant.chars, ...chars};
```
which would mean you are copying and recreating the object every time `defineChars()` is called.  Adding the new characters into the existing object is more efficient, and is required in order to keep the inheritance intact.